File: Balanced Binary Tree - Leetcode 110.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        balanced = [True]

        def height(root):
            if not root:
                return 0

            left_height = height(root.left)
            if balanced[0] is False:
                return 0


            right_height = height(root.right)
            if abs(left_height - right_height) > 1:
                balanced[0] = False
                return 0
            return 1 + max(left_height, right_height)

        height(root)
        return balanced[0]

# Time Complexity: O(n)
# Space Complexity: O(h) { here "h" is the height of the tree }

File: Balanced Binary Tree - Leetcode 110/Balanced Binary Tree - Leetcode 110.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = val;
 *     this.left = left === undefined ? null : left;
 *     this.right = right === undefined ? null : right;
 * }
 */

var isBalanced = function(root) {
    return height(root) !== -1;
};

function height(node) {
    if (node === null) return 0;

    const leftHeight = height(node.left);
    if (leftHeight === -1) return -1;

    const rightHeight = height(node.right);
    if (rightHeight === -1) return -1;

    if (Math.abs(leftHeight - rightHeight) > 1) {
        return -1;
    }

    return 1 + Math.max(leftHeight, rightHeight);
}
